### PR TITLE
kubectl: auto correction for typos in resource name

### PR DESCRIPTION
**Pull request by anfernee**
When typing the wrong resource type name, make kubectl smart enough to
recommend the true resource name based on the shorted edit distance.
Inspired by `git`.

$ ./kubectl get nodex
the server doesn't have a resource type "nodex"

Did you mean one of these?

	nodes

$ ./kubectl get deplyo.extensions
the server doesn't have a resource type "deplyo" in group "extensions"

Did you mean one of these?

	deploy.extensions

**What this PR does / why we need it**: more user-friendly.

**Release note**:

```release-note
kubectl: auto correction for typos in subcommand and resource name
```

### CODE REVIEW

**Comment by anfernee** [comment]
cc @kubernetes/sig-cli-maintainers

File: staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go [hunk 1]
```diff
@@ -51,7 +51,7 @@ func (gr GroupResource) Empty() bool {
 	return len(gr.Group) == 0 && len(gr.Resource) == 0
 }
 
-func (gr *GroupResource) String() string {
+func (gr GroupResource) String() string {
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go]: Why this change? It makes this type inconsistent with the other types in this file.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 2]
```diff
@@ -136,3 +159,108 @@ func (e shortcutExpander) expandResourceShortcut(resource schema.GroupVersionRes
...
+	allNames := map[string]bool{}
+	for _, shortcut := range e.cachedShortcuts {
+		allNames[shortcut.LongForm.String()] = true
+		allNames[shortcut.ShortForm.String()] = true
+		allNames[shortcut.LongForm.Resource] = true
+		allNames[shortcut.ShortForm.Resource] = true
+	}
+
+	results := map[int][]string{}
+	res := resource.GroupResource().String()
```
Comment (anfernee) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: here it is.

File: staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go (hunk 1 above)
Comment (anfernee) [reply staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go]: Thanks for review @liggitt. The reason is I wanted to do this: `resource.GroupResource().String()`. and  `resource.GroupResource()` doesn't return a pointer. It's weird, but I can do it without this change.

Good catch. I will change it back.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 2 above)
Comment (anfernee) [reply pkg/kubectl/cmd/util/shortcut_restmapper.go]: i changed it to this:

```
       gr := resource.GroupResource()
       res := (&gr).String()
```

a little bit awkward.

File: staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go (hunk 1 above)
Comment (anfernee) [reply staging/src/k8s.io/apimachinery/pkg/runtime/schema/group_version.go]: I reverted the change. The point of using pointer as receiver is probably to avoid copying values, but the previous methods use values receivers and return values. so it's confusing. however, it's out of the scope of this change.

File: pkg/kubectl/cmd/util/helpers.go [hunk 3]
```diff
@@ -147,17 +147,29 @@ func checkErr(err error, handleErr func(string, int)) {
...
+				msg = fmt.Sprintf("the server doesn't have a resource type %q in version %q", err.PartialResource.Resource, err.PartialResource.Version)
 			default:
-				handleErr(fmt.Sprintf("the server doesn't have a resource type %q", err.PartialResource.Resource), DefaultErrorExitCode)
+				msg = fmt.Sprintf("the server doesn't have a resource type %q", err.PartialResource.Resource)
+			}
+
+			if len(err.Recommends) != 0 {
+				msg += "\n\nDid you mean one of these?\n\n"
+				for _, recommend := range err.Recommends {
+					msg += fmt.Sprintf("\t%s\n", recommend)
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/helpers.go]: Are we indenting with tab elsewhere, or is this the only place?

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 4]
```diff
@@ -136,3 +159,109 @@ func (e shortcutExpander) expandResourceShortcut(resource schema.GroupVersionRes
...
+	for i := 1; i <= max; i++ {
+		if result, ok := results[i]; ok {
+			return result
+		}
+	}
+
+	return []string{}
+}
+
+func editDistance(s1, s2 string) int {
```
Comment (smarterclayton) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: Don’t we already have a dependency on an edit distance library via spf13/cobra, which already suggests this for command names?  If so, use that library.

**Review by smarterclayton — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/util/helpers.go (hunk 3 above)
Comment (anfernee) [reply pkg/kubectl/cmd/util/helpers.go]: Thanks @smarterclayton for reviewing. 

The tabs are indenting for recommended commands. I think it would help. but I am open to options, like 2 spaces, 4 spaces, '- xxx', '* xxx' . What do you suggest?

```console
$ kubectl bet
Error: unknown command "bet" for "kubectl"

Did you mean this?
	set
	get

Run 'kubectl --help' for usage.
$ git bet
git: 'bet' is not a git command. See 'git --help'.

Did you mean this?
	bisect
```

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 4 above)
Comment (anfernee) [reply pkg/kubectl/cmd/util/shortcut_restmapper.go]: @smarterclayton yes, kubectl is using cobra. to my best knowledge, cobra deals with predefined subcommands, and suggests names for wrong subcommands. this patch also suggests wrong resource names, which are dynamic. They have a levenshteinDistance method, but it's not intended for consumption, also it's private.

**Comment by shiywang** [comment]
@anfernee could you change the release note to a more concrete description ? also tittle, thx

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 5]
```diff
@@ -27,59 +27,80 @@ import (
 	"k8s.io/kubernetes/pkg/kubectl"
 )
 
-// shortcutExpander is a RESTMapper that can be used for Kubernetes resources.   It expands the resource first, then invokes the wrapped
+// NoResourceWithCommendError
+type NoResourceWithCommendError struct {
+	*meta.NoResourceMatchError
+
+	Recommends []string
```
Comment (soltysh) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: `Suggestions`

File: pkg/kubectl/cmd/util/helpers.go (hunk 3 above)
Comment (soltysh) [reply pkg/kubectl/cmd/util/helpers.go]: This is actually coming from cobra.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 4 above)
Comment (soltysh) [reply pkg/kubectl/cmd/util/shortcut_restmapper.go]: Can you also move it to a separate file (`helpers.go`) , please?

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 6]
```diff
@@ -99,14 +120,16 @@ func (e shortcutExpander) getShortcutMappings() ([]kubectl.ResourceShortcuts, er
 
 	// append hardcoded short forms at the end of the list
 	res = append(res, kubectl.ResourcesShortcutStatic...)
+	e.cachedShortcuts = res
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: this isn't threadsafe

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 7]
```diff
@@ -136,3 +159,109 @@ func (e shortcutExpander) expandResourceShortcut(resource schema.GroupVersionRes
 
 	return resource
 }
+
+func (e *shortcutExpander) wrapError(err error, resource schema.GroupVersionResource) error {
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: I wouldn't expect this in the shortcutExpander, but in a separate helper

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 8]
```diff
@@ -136,3 +159,109 @@ func (e shortcutExpander) expandResourceShortcut(resource schema.GroupVersionRes
...
+		return &NoResourceWithCommendError{
+			NoResourceMatchError: er,
+			Recommends:           e.similarResources(resource),
+		}
+	}
+
+	return err
+}
+
+func (e *shortcutExpander) similarResources(resource schema.GroupVersionResource) []string {
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: this is separate function... make it a separate component that takes either a discovery interface or a list of resources, along with the specified resource, and makes suggestions

File: pkg/kubectl/cmd/util/shortcut_restmapper.go [hunk 9]
```diff
@@ -27,59 +27,80 @@ import (
...
 
 	discoveryClient discovery.DiscoveryInterface
+	cachedShortcuts []kubectl.ResourceShortcuts
 }
 
 var _ meta.RESTMapper = &shortcutExpander{}
 
-func NewShortcutExpander(delegate meta.RESTMapper, client discovery.DiscoveryInterface) shortcutExpander {
-	return shortcutExpander{RESTMapper: delegate, discoveryClient: client}
+func NewShortcutExpander(delegate meta.RESTMapper, client discovery.DiscoveryInterface) *shortcutExpander {
```
Comment (liggitt) [inline pkg/kubectl/cmd/util/shortcut_restmapper.go]: I wouldn't expect changes to shortcut expander... suggestions are useful even if you're not using expansion... separate out the suggestion function

**Comment by anfernee** [comment]
Thanks guys for your comments. another run :)

**Comment by anfernee** [comment]
ping @soltysh @smarterclayton @liggitt

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 9 above)
Comment (apelisse) [reply pkg/kubectl/cmd/util/shortcut_restmapper.go]: @anfernee Did you address that comment?

File: pkg/kubectl/cmd/util/suggestion_restmapper.go [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+func (m *suggestionRESTMapper) RESTMapping(gk schema.GroupKind, versions ...string) (*meta.RESTMapping, error) {
+	return m.delegate.RESTMapping(gk, versions...)
+}
+
+func (m *suggestionRESTMapper) RESTMappings(gk schema.GroupKind, versions ...string) ([]*meta.RESTMapping, error) {
+	return m.delegate.RESTMappings(gk, versions...)
+}
+
+func (m *suggestionRESTMapper) similarResources(resource schema.GroupVersionResource) []string {
```
Comment (droot) [inline pkg/kubectl/cmd/util/suggestion_restmapper.go]: One of the ways to isolate this functionality will be having an independent function `similarResources` which takes in `[]kubectl.ResourceShortcuts` and a `schema.GroupVersionResource` and returns similar resources, so it assumes ResourceShortcuts are passed to it.
And then we can don't have to extend ShortcutExpander, instead just invoke this new function `similarResources` in the NoResourceMatchError error handing.
I think that will address @liggitt concerns about the tight-coupling of this with shortcutexpander or he might have better suggestion about it.

**Review by droot — COMMENTED** [review]
Very useful functionality. Took a look, have a minor suggestion about code-reorg. Happy to discuss offline as well.

File: pkg/kubectl/cmd/util/shortcut_restmapper.go (hunk 9 above)
Comment (anfernee) [reply pkg/kubectl/cmd/util/shortcut_restmapper.go]: yes, i did.

File: pkg/kubectl/cmd/util/suggestion_restmapper.go (hunk 10 above)
Comment (anfernee) [reply pkg/kubectl/cmd/util/suggestion_restmapper.go]: thanks @droot for the review. It makes sense. Now I wrapped the suggestion related functionality in side this `suggestionRESTMapper`, so it's already decoupled with shortcut expander.

File: pkg/kubectl/cmd/util/helpers.go [hunk 11]
```diff
@@ -834,3 +849,121 @@ func ShouldIncludeUninitialized(cmd *cobra.Command, includeUninitialized bool) b
...
+// getShortcutsWithCache is the same as getShortcutMappings, but it caches result.
+func (d *resourceDiscover) getShortcutsWithCache() ([]kubectl.ResourceShortcuts, error) {
+	d.lock.Lock()
+	defer d.lock.Unlock()
+
+	if d.resources != nil {
+		return d.resources, nil
+	}
+
+	resources, err := d.getShortcutMappings()
```
[the remark this answers is not in the document]
Comment (anfernee) [reply pkg/kubectl/cmd/util/helpers.go]: Good catch. it should be nil for error, but it shouldn't.

[the remark this answers is not in the document]
Comment (anfernee) [reply pkg/kubectl/cmd/util/helpers.go]: I removed the cache, because http client is already cached results.

File: pkg/kubectl/cmd/util/suggestion_restmapper.go (hunk 10 above)
Comment (soltysh) [reply pkg/kubectl/cmd/util/suggestion_restmapper.go]: I think this and other Jordan's comment still isn't addressed.

**Review by soltysh — CHANGES_REQUESTED** [review]

File: pkg/kubectl/cmd/util/suggestion_restmapper.go (hunk 10 above)
Comment (anfernee) [reply pkg/kubectl/cmd/util/suggestion_restmapper.go]: I must mis-understand something here. I re-read @liggitt 's comment. @soltysh can you kindly point out which one it doesn't address? I can only think of wrapping line 87-122 to a separate function? If it's about decouping `ShortcutExpander`, it's already decoupled, both `ShortcutExpander` and `suggestionRESTMapper` are decoration to `RESSTMapper`.  suggestion knows nothing about expander. you can basically enable/disable either suggestion or expander separately with this PR.
